### PR TITLE
SQL-627: Implement SQLGetEnvAttr and SQLSetEnvAttr

### PR DESCRIPTION
**Pull request by varsha804**

### CODE REVIEW

File: odbc/src/api.rs [hunk 1]
```diff
@@ -800,13 +801,40 @@ pub extern "C" fn SQLGetEnvAttr(
 
 #[no_mangle]
 pub extern "C" fn SQLGetEnvAttrW(
-    _environment_handle: HEnv,
-    _attribute: EnvironmentAttribute,
-    _value_ptr: Pointer,
+    environment_handle: HEnv,
+    attribute: EnvironmentAttribute,
+    value_ptr: Pointer,
     _buffer_length: Integer,
```
Comment (varsha804) [inline odbc/src/api.rs:807]: We don't need to use `buffer_length` since none of the environment attributes are character strings.

File: odbc/src/api.rs [hunk 2]
```diff
@@ -1115,12 +1143,56 @@ pub extern "C" fn SQLSetEnvAttr(
...
 ) -> SqlReturn {
-    unimplemented!()
+    // let mut env_handle = environment_handle as *mut MongoHandle;
+    let env_handle = MongoHandleRef::from(environment_handle);
+    env_handle.clear_diagnostics();
+    match env_handle.as_env() {
+        None => SqlReturn::INVALID_HANDLE,
+        Some(env) => match attribute {
+            EnvironmentAttribute::OdbcVersion => match value as Integer {
+                380 => {
```
Comment (varsha804) [inline odbc/src/api.rs:1158]: Ideally, I'd be able to match on the `odbc-sys` enums (e.g. `AttrOdbcVersion`), but those enums don't derive `FromPrimitive`.

File: odbc/src/tests.rs [hunk 3]
```diff
@@ -463,3 +465,326 @@ mod get_diag_rec {
...
+    };
+    use std::{ffi::c_void, sync::RwLock};
+
+    const INVALID_ATTR_VALUE: &str = "HY024\0";
+
+    #[test]
+    fn get_set_odbc_version() {
+        let env_handle: *mut _ =
+            &mut MongoHandle::Env(RwLock::new(Env::with_state(EnvState::Allocated)));
+        // SQLSetEnvAttrW(SQL_ATTR_ODBC_VERSION, SQL_OV_ODBC3_80)
```
Comment (varsha804) [inline odbc/src/tests.rs:486]: The comments state the obvious, but they at least helped me to quickly get the gist of each function

**Review by pmeredit — CHANGES_REQUESTED** [review]
Overall this looks great! Just a few small questions

File: odbc/src/api.rs [hunk 4]
```diff
@@ -800,13 +801,41 @@ pub extern "C" fn SQLGetEnvAttr(
...
+        None => SqlReturn::INVALID_HANDLE,
+        Some(env) => {
+            let env_contents = env.read().unwrap();
+            unsafe {
+                *string_length = match value_ptr.is_null() {
+                    true => 0,
+                    false => size_of::<Integer>() as Integer,
+                }
+            }
+            match attribute {
```
Comment (nbagnard) [inline odbc/src/api.rs:822]: Are you sure that this will not blow up if value_ptr is null? Is there some kind of Rust magic?

Also, value_ptr is a pointer to a buffer the user/app provides to store the output data. If a null pointer is passed, how can we can change it. If we were to provide the pointer to the output buffer it would work, but it is not the case.

File: odbc/src/api.rs [hunk 5]
```diff
@@ -1115,12 +1144,55 @@ pub extern "C" fn SQLSetEnvAttr(
...
+        None => SqlReturn::INVALID_HANDLE,
+        Some(env) => match attribute {
+            EnvironmentAttribute::OdbcVersion => match value as Integer {
+                380 => {
+                    let mut env_contents = (*env).write().unwrap();
+                    env_contents.attributes.odbc_ver = value as _;
+                    SqlReturn::SUCCESS
+                }
+                _ => {
+                    env_handle.add_diag_info(ODBCError::InvalidAttrValue("SQL_ATTR_ODBC_VERSION"));
```
Comment (nbagnard) [inline odbc/src/api.rs:1164]: I don't know if this is compliant. This is for the application to set the version of ODBC it's implemented against. It is a little different from the version that the driver supports which is returned via SqlGetInfo.
I think that the driver is supposed to implement a different behavior based on what the application sets.
But, because I don't want to have to deal with duplicating the error codes (odbc 2 and 3 error codes are different) and slight changes in the odbc state transition. 
If everyone agrees, let's try this and see if users are upset.

File: odbc/src/api.rs [hunk 6]
```diff
@@ -1115,12 +1144,55 @@ pub extern "C" fn SQLSetEnvAttr(
...
+                    SqlReturn::ERROR
+                }
+            },
+            EnvironmentAttribute::ConnectionPooling => match value as Integer {
```
Comment (nbagnard) [inline odbc/src/api.rs:1168]: umh, supposedly "SQL_ATTR_CONNECTION_POOLING is implemented inside the Driver Manager.", so we could just ignore whatever value is passer maybe?
PowerBi will set this to "SQL_CP_ONE_PER_HENV" so better not fails there or we won't be able to do anything :D
I don't think that the call will make to the driver, so it might be OK. We'll find out when we have a simple connection implemented.

File: odbc/src/api.rs [hunk 7]
```diff
@@ -1115,12 +1144,55 @@ pub extern "C" fn SQLSetEnvAttr(
...
+                0 => SqlReturn::SUCCESS,
+                _ => {
+                    env_handle
+                        .add_diag_info(ODBCError::Unimplemented("SQL_ATTR_CONNECTION_POOLING"));
+                    SqlReturn::ERROR
+                }
+            },
+            EnvironmentAttribute::CpMatch => match value as Integer {
```
Comment (nbagnard) [inline odbc/src/api.rs:1176]: Since we don't support pooling, I think that none of the option are valid, no?
But again, this looks like a DM related setting, so maybe it's just deadcode.

File: odbc/src/api.rs (hunk 2 above)
Comment (nbagnard) [reply odbc/src/api.rs:1158]: oh :(
Could you do something like getting a int value from the enum though? So that you don't have to hardcode the possible values and we have less room for mistakes (and if there mistake it will be from odbc-sys)?

File: odbc/src/api.rs [hunk 8]
```diff
@@ -1115,12 +1144,55 @@ pub extern "C" fn SQLSetEnvAttr(
...
+                0 => SqlReturn::SUCCESS,
+                _ => {
+                    env_handle.add_diag_info(ODBCError::Unimplemented("SQL_CP_RELAXED_MATCH"));
+                    SqlReturn::ERROR
+                }
+            },
+            EnvironmentAttribute::OutputNts => match unsafe { *(value as *const Integer) } {
+                1 => SqlReturn::SUCCESS,
+                0 => {
```
[the remark this answers is not in the document]
Comment (nbagnard) [reply odbc/src/api.rs:1185]: That'd be nice.

File: odbc/src/api.rs [hunk 9]
```diff
@@ -800,13 +801,41 @@ pub extern "C" fn SQLGetEnvAttr(
...
+) -> SqlReturn {
+    let env_handle = MongoHandleRef::from(environment_handle);
+    env_handle.clear_diagnostics();
+    match env_handle.as_env() {
+        None => SqlReturn::INVALID_HANDLE,
+        Some(env) => {
+            let env_contents = env.read().unwrap();
+            unsafe {
+                *string_length = match value_ptr.is_null() {
+                    true => 0,
```
[the remark this answers is not in the document]
Comment (varsha804) [reply odbc/src/api.rs:818]: Ah yes, I read this sentence a few times when figuring out what to set `string_length` to when `value_ptr` is null:

> If ValuePtr is NULL, StringLengthPtr will still return the total number of bytes (excluding the null-termination character for character data) available to return in the buffer pointed to by ValuePtr.

and wasn't sure if "the total number of bytes [...] available to return in the buffer" means the size of the buffer or the length of the attribute's value -- not sure how exactly to interpret "available to return" 😅 . I went with option one, but I can switch to option two if that's what you think they're saying.

> I tried the function with OdbcTest

I wasn't aware of ODBC Test until now! I'll try it out

> Even though the spec does not flag it as such, the DM takes care of HY092 (good since I did not see how this could be otherwise).

So we don't need to log a diagnostic record when an attribute value is invalid?

File: odbc/src/api.rs (hunk 4 above)
Comment (varsha804) [reply odbc/src/api.rs:822]: Ah, looks like I'm missing a null pointer check! I'll add that in

File: odbc/src/api.rs (hunk 6 above)
Comment (varsha804) [reply odbc/src/api.rs:1168]: Oh interesting, maybe I can change it so that we only accept `SQL_CP_ONE_PER_HENV` instead of `SQL_CP_OFF`

File: odbc/src/api.rs (hunk 2 above)
Comment (varsha804) [reply odbc/src/api.rs:1158]: I could do `if value as Integer == AttrOdbcVersion::Odbc3_80 as Integer { ... } else { ... }`

**Review by nbagnard — CHANGES_REQUESTED** [review]
Except the handling and testing of Null ptr for ValuePointer, things look good!
Everything else are small spec details and things we might have to update when we have a connection integrated.

File: odbc/src/tests.rs [hunk 10]
```diff
@@ -463,3 +465,326 @@ mod get_diag_rec {
         );
     }
 }
+mod env_attributes {
```
[the remark this answers is not in the document]
Comment (varsha804) [reply odbc/src/tests.rs:468]: > Can you add tests with ValuePtr being Null for all environment attributes ?

Done.

> should we add tests where ValuePtr is an invalid value

I don't think this case needs to be tested anymore now that I have an enum for every environment attribute. Typecasting an invalid value to an environment attribute enum variant will result in a successful cast instead of an error, so I think just testing the individual variants should be enough.

File: odbc/src/api.rs (hunk 5 above)
Comment (varsha804) [reply odbc/src/api.rs:1164]: That makes sense to me. Just updated this part of the function

File: odbc/src/api.rs (hunk 6 above)
Comment (varsha804) [reply odbc/src/api.rs:1168]: Filed [SQL-681 ](https://jira.mongodb.org/browse/SQL-681)so we can investigate this further

File: odbc/src/api.rs (hunk 5 above)
Comment (pmeredit) [reply odbc/src/api.rs:1166]: @nbagnard that sounds good to me. Nothing lower than 3.0 though.

**Review by pmeredit — APPROVED** [review]
LGTM!

File: odbc/src/tests.rs [hunk 11]
```diff
@@ -463,3 +465,401 @@ mod get_diag_rec {
...
+    use odbc_sys::{
+        AttrConnectionPooling, AttrCpMatch, AttrOdbcVersion, EnvironmentAttribute, HEnv,
+        HandleType, Pointer, SqlReturn,
+    };
+    use std::{ffi::c_void, sync::RwLock};
+
+    const OPTIONAL_VALUE_CHANGED: &str = "01S02\0";
+
+    #[test]
+    fn get_set_odbc_version() {
```
Comment (nbagnard) [inline odbc/src/tests.rs:484]: Have you tried to write a helper for these tests?
The logic (and most of the code) is the same for all of them.
Having a helper would:
- Allow you to have a concise and logical "test" that can be applied to all the attributes and this way you can make sure the same things are tested for all of them and nothing is forgotten in one of the [test].
- Make things easier when you add a new test case (like testing string_length is null).

File: odbc/src/tests.rs [hunk 12]
```diff
@@ -463,3 +465,401 @@ mod get_diag_rec {
...
+        );
+        assert_eq!(OPTIONAL_VALUE_CHANGED, unsafe {
+            String::from_utf16(&*(sql_state as *const [u16; 6])).unwrap()
+        });
+        assert_eq!(
+            "[MongoDB][API] Invalid value for attribute SQL_ATTR_CONNECTION_POOLING\0",
+            unsafe { String::from_utf16(&*(message_text as *const [u16; 71])).unwrap() }
+        );
+        // Exclude the number of characters required for the null terminator
+        assert_eq!(70, *text_length_ptr);
```
Comment (nbagnard) [inline odbc/src/tests.rs:741]: Keeping the assert on the error code and error message is great because we want to make sure that we get the "right error, but I don't think you need to add these 2 asserts. You already added tests for SQLGetDiagRec and they cover this.

**Review by nbagnard — CHANGES_REQUESTED** [review]
Only a few more comments. It is still looking good :D 
The main thing being what is likely a possible NPE in SQLGetEnvAttr. 
The others are more related to tests coverage and "layout".

File: odbc/src/tests.rs (hunk 11 above)
Comment (varsha804) [reply odbc/src/tests.rs:484]: Done -- it's a little odd because I had to convert every attribute into an `i32` in order to use a helper that works for any environment attribute, but it does make the testing code much cleaner!

File: odbc/src/api.rs [hunk 13]
```diff
@@ -800,13 +803,41 @@ pub extern "C" fn SQLGetEnvAttr(
...
+    let env_handle = MongoHandleRef::from(environment_handle);
+    env_handle.clear_diagnostics();
+    match env_handle.as_env() {
+        None => SqlReturn::INVALID_HANDLE,
+        Some(env) => {
+            let env_contents = env.read().unwrap();
+            if value_ptr.is_null() {
+                unsafe { *string_length = 0 }
+            } else {
+                unsafe { *string_length = size_of::<Integer>() as Integer }
```
[the remark this answers is not in the document]
Comment (varsha804) [reply odbc/src/api.rs:821]: Oh good catch -- added a test for this

File: odbc/src/tests.rs [hunk 14]
```diff
@@ -463,3 +465,401 @@ mod get_diag_rec {
         );
     }
 }
+mod env_attributes {
```
[the remark this answers is not in the document]
Comment (varsha804) [reply odbc/src/tests.rs:468]: That would be nice, but I don't think we can do that with our current file structure(Rust [docs](https://doc.rust-lang.org/book/ch11-03-test-organization.html)). I tried creating a separate test file for the environment attributes tests, but `cargo test` wasn't able to find it.

[the remark this answers is not in the document]
Comment (nbagnard) [reply odbc/src/tests.rs:468]: You can separate them.
You need to have the corresponding module declarations annoted with #[cfg(test)].

You can add the test modules declarations in lib.rs and have the file at the same level than the code itself. Or you can put it in api.rs and have the tests under src/api. I've tested it and both work.
Like this and then the code for the related tests goes in the corresponding .rs file.
```
#[cfg(test)]
mod handles_tests;

#[cfg(test)]
mod get_diag_rec_tests;
```

File: odbc/src/tests.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (nbagnard) [reply odbc/src/tests.rs:468]: Awesome work, I like this new environment attributes test layout a lot!

File: odbc/src/tests.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (varsha804) [reply odbc/src/tests.rs]: Oh nice, not sure why I was running into troubles earlier 🤔 I decided to create separate `api` and `handles` directories, and have removed `src/tests.rs`.

**Review by nbagnard — APPROVED** [review]
LGTM! Thank you :)

**Comment by varsha804** [comment]
evergreen merge
